Fix for issue #491

thewindsofwinter opened the pull request:
## Summary: a response to #491 (clock module crash)

### Issue Identification
Certain sections of the wikipedia page for timezones do not have a link for the time zone (for example, the first two rows displayed in the image below).

![timezones](https://user-images.githubusercontent.com/46852717/70487907-735efc00-1abc-11ea-8bc7-ea7cb613f122.PNG)

### Resolution
After consulting the guide for HTML scraping at http://effbot.org/zone/element-index.htm#documentation, I added a failsafe for the cases in which there is no link. Since the footnotes are made using `<sub>`, and the `.text` function returns the string before any sub-elements, this should fix the issue mentioned above.

Make sure to test if this works.

mr-martian commented:
Looking at the tests, it looks like the same problem shows up again at line 193, which is doing the same thing with the UTC Offset column.

thewindsofwinter commented:
Thanks @mr-martian, I just pushed again to fix that problem :)

thewindsofwinter commented:
Okay, looks like all build tests involving clock.py are succeeding, for documentation purposes I fixed another bug around Line 234. Basically I updated the column selector from "TZ" to the actual "TZ Database Name" and I updated "UTC Offset" to "UTC Offset ±hh:mm" in accordance with the column changes.

![changetz](https://user-images.githubusercontent.com/46852717/70583093-39f0c400-1b82-11ea-9b2f-676fcc10a7a5.PNG)

wei2912 commented:
@thewindsofwinter to check, by any chance are you a current GCI student?

thewindsofwinter commented:
Yeah, @wei2912

wei2912 commented:
Nice, if you fix the clock tests too I reckon we could approve this for a GCI task - will draft up the task description and let you know how to claim it.

mr-martian commented:
A task for this actually already exists and was approved.

wei2912 commented:
Oh, I must have missed that out. Thanks for the clarification!

Androbin commented on modules/clock.py:
```diff
@@ -180,9 +180,26 @@ def scrape_wiki_time_zone_abbreviations(doc):
+                # to fix through testing whether cell has
+                # a link inside of it, otherwise just get text.
+
+                if len(cell.findall('a')) == 0:
```
Can we factor this into a dedicated function?

thewindsofwinter replied:
Understandable. I can do that shortly.

Androbin commented on modules/clock.py:
```diff
@@ -219,12 +236,14 @@ def scrape_wiki_tz_database_time_zones(doc):
         for cell in row.findall('td'):
-            if column == column_names.index('TZ'):
+            # issue with finding the column, fixing
+            if column == column_names.index('TZ database name'):
```
Maybe it would be more robust / future-proof to use a regex `(^| )TZ( |$)` or something?

thewindsofwinter replied:
I could, as for right now I was worried by introducing another level of complexity using regular expressions. It would be even more robust to simply capture an image of the webpage and use that as a more static format than the actual wikipedia page which can be edited at a moment's notice.

Thanks for reviewing,
Andy

Androbin replied:
We do have a caching mechanism in place, but the point of this thing is to be up-to-date.

thewindsofwinter replied:
@Androbin I've made a new commit to do the things that you asked for.

Androbin commented on modules/clock.py:
```diff
@@ -25,6 +25,18 @@
+    if len(cell.findall('a')) == 0:
+        offset = cell.text[3:]
+    else:
+        offset = cell.find('a').text[3:]
```
It is not obvious: Why are the first 3 characters stripped?
Also, please find the right cell/text and do further processing in two steps, so you don't have to duplicate logic.

thewindsofwinter replied:
Thanks for catching that! It was part of a different processing where the "UTC" is removed from offset, but it shouldn't be in the main function. Fixing

Androbin commented on modules/clock.py:
```diff
@@ -219,12 +234,16 @@ def scrape_wiki_tz_database_time_zones(doc):
+            # issue with finding the column, fixing
+            # Used regex as mentioned in the pull-request review
+            current_regex = re.compile("(^| )TZ( |$)")
+            if column == column_names.index(list(filter(current_regex.match, column_names))[0]):
```
No need to do `list(filter(current_regex.match, column_names))[0]`
if you can just do `next(filter(current_regex.match, column_names)`

thewindsofwinter replied:
Thanks, I'm not fully familiar with every function of the Python language. Will include!

Androbin commented on modules/clock.py:
```diff
@@ -219,12 +234,16 @@ def scrape_wiki_tz_database_time_zones(doc):
                 name = text.split('/')[-1]
-            elif column == column_names.index('UTC offset'):
+            # used to be UTC offset but I think complete names are necessary
+            elif column == column_names.index('UTC offset ±hh:mm'):
```
Please use the same regex logic as above. Maybe refactor into a dedicated function.

thewindsofwinter replied:
I will do that soon.

thewindsofwinter replied:
Updated things, hopefully the code is better now.

Androbin commented:
LGTM @wei2912

wei2912 commented:
@thewindsofwinter @Androbin Thanks for the merge!
